nixos/modules/services/networking/iwd.nix

```diff
@@ -4,16 +4,142 @@ with lib;
 
 let
   cfg = config.networking.wireless.iwd;
+
+  encodeFileName = pkgs.writeScript "encode" ''
+    #! ${pkgs.runtimeShell} -e
+    echo "=$(printf "$1" | ${pkgs.unixtools.xxd}/bin/xxd -pu)"
+  '';
+
+  # If a network consists of alphanumeric chars and `_` and `-`, the network will be stored
+  # as `/var/lib/iwd/<ssid>.<ext>`. In any other case a hex-encoded version of the SSID will
+  # be stored with an `=` as prefix.
+  encodeIfNeeded = file: ext: let
+    allowedChars = stringToCharacters "01234556789abcdefghijklmnopqrstuvwxyz-_ ";
+    skipEncoding = all (char: elem (toLower char) allowedChars) (stringToCharacters file);
+  in
+    "/var/lib/iwd/${if skipEncoding then file else "$(${encodeFileName} \"${file}\")"}.${ext}";
+
+  # Basically `types.str`, but the `check` function validates the string-length as well.
+  fixedStringLength = min: max: mkOptionType {
+    name = "str-with-length";
+    description = if min == max
+      then "string with exactly ${toString min} chars."
+      else "string which must have at least ${toString min} and at most ${toString max} chars.";
+    inherit (types.str) merge;
+    check = x: isString x && stringLength x >= min && stringLength x <= max;
+  };
+
+  # The file extension of the network-config determines the type of the network. Currently,
+  # we support public networks (`open`) and networks protected with a passphrase (`psk`).
+  # Anything else, like 802.1x networks are not supported by this module.
+  determineType = cfg: if cfg.psk != null || cfg.passphrase != null then "psk" else "open";
+
+  # Generates a name-value-pair with configurations for a SSID. For further information
+  # about the configuration options, please refer to the upstream docs:
+  #
+  # https://iwd.wiki.kernel.org/networkconfigurationsettings
+  predefinedNetworks = flip mapAttrs cfg.networks (name: cfg: {
+    extension = determineType cfg;
+
+    # Cannot use the SSID here: the hex-encode is done at build-time, not eval time
+    # and several special chars in a file name would break `writeText`.
+    file = pkgs.writeText "ssid" (generators.toINI {} ({
+      Settings.AutoConnect = cfg.autoConnect;
```

Comment:
You can also do this at runtime by dropping the use of `generators.toINI` and manually constructing the INI file

Reply:
Well, I'm not sure whether we should do this: I check if a file in `/var/lib/iwd` is a store-path and if that's the case, IWD now refuses to modify/remove/whatever this network configuration. This wouldn't be possible anymore with runtime-construction and thus leading to eventual consistency as `iwd` can simply modify the network's configuration and therefore breaking the immutability-goal of NixOS.

Reply:
It's still possible to do this, which imo is even nicer since it doesn't rely on checking for Nix store symlinks: Write the list of files that were deployed declaratively to `/var/lib/iwd/.declarative-nixos-networks` after linking them. Then in the next `preStart`, remove all the files given in that list first.

Reply:
I like the idea and will give it a try soon :)